X16 Reference - Appendix F - 65C816 Processor.md

```diff
@@ -0,0 +1,2517 @@
+wise to clear **m** and **x** with `REP #$30`.
+
+* Set .X to the source_address + size - 1
+* Set .Y to the destination_address
+* Set .A to size-1
+* MVP #source_bank, #dest_bank
+
+
+[[Opcodes](#instructions-by-opcode)] [[By Name](#instructions-by-name)] [[By Category](#instructions-by-category)]
+
+---
+
+### NOP
+
+**No Operation**
+
+```text
+SYNTAX           MODE       HEX LEN CYCLES      FLAGS   
+NOP              imp        EA  1   2           ........ .
+```
+
+The CPU performs no operation. This is useful when blocking out instructions, or
+reserving space for later use.
+
+
+[[Opcodes](#instructions-by-opcode)] [[By Name](#instructions-by-name)] [[By Category](#instructions-by-category)]
+
+---
+
+### ORA
+
+**Boolean OR**
+
+```text
+SYNTAX           MODE       HEX LEN CYCLES      FLAGS   
+ORA #$20/#$1234  imm        09  3-m 3-m         n.....z. .
+ORA $20          dir        05  2   4-m+w       n.....z. .
+ORA $20,X        dir,X      15  2   5-m+w       n.....z. .
+ORA $20,S        stk,S      03  2   5-m         n.....z. .
+ORA $1234        abs        0D  3   5-m         n.....z. .
+ORA $1234,X      abs,X      1D  3   6-m-x+x*p   n.....z. .
+ORA $1234,Y      abs,Y      19  3   6-m-x+x*p   n.....z. .
+ORA $FEDBCA      long       0F  4   6-m         n.....z. .
+ORA $FEDCBA,X    long,X     1F  4   6-m         n.....z. .
+ORA ($20)        (dir)      12  2   6-m+w       n.....z. .
+ORA ($20),Y      (dir),Y    11  2   7-m+w-x+x*p n.....z. .
+ORA ($20,X)      (dir,X)    01  2   7-m+w       n.....z. .
+ORA ($20,S),Y    (stk,S),Y  13  2   8-m         n.....z. .
+ORA [$20]        [dir]      07  2   7-m+w       n.....z. .
+ORA [$20],Y      [dir],Y    17  2   7-m+w       n.....z. .
+```
+
+Perform a Boolean OR operation with the operand and .A
+
+ORA compares each bit of the operands and sets the result bit to 1 if either or
+both of the two bits is 1. If both bits are 0, the result is 0.
+
+ORA is useful for *setting* a specific bit in a byte.
+
+Truth table for ORA:
+
+```text
+Operand 1: 1100
+Operand 2: 1010
+Result:    1110
+```
+
+
+[[Opcodes](#instructions-by-opcode)] [[By Name](#instructions-by-name)] [[By Category](#instructions-by-category)]
+
+---
+
+### PEA
+
+**Push Absolute**
+
+```text
+SYNTAX           MODE       HEX LEN CYCLES      FLAGS   
+PEA $2034        imm        F4  3   5           ........ .
+```
+
+PEA, PEI, and PER push values to the stack *without* affecting registers.
+
+PEA pushes the operand value onto the stack. The literal operand is used, rather
+than an address.
+
+Some assemblers might not assemble `PEA $1234`. If you run into this problem,
+use the immediate mode syntax `PEA #$1234`.
+
+
+[[Opcodes](#instructions-by-opcode)] [[By Name](#instructions-by-name)] [[By Category](#instructions-by-category)]
+
+---
+
+### PEI
+
+**Push Effecive Indirect Address**
+
+```text
+SYNTAX           MODE       HEX LEN CYCLES      FLAGS   
+PEI ($20)        dir        D4  2   6+w         ........ .
+```
```

Comment:
does the mode need to be changed here too?  Is it (dir)?

Reply:
Technically, that's Direct Indirect. I'll look at how WDC documents it in the datasheet.

Reply:
PER/PEI/PEA all purport to push addresses to the stack, even though they don't have to be addresses for PEI and PEA. I think because it's treating the value inside of direct page as an address, even though it doesn't read from it, the (addr) convention is used

Reply:
Ok, so I guess... do we want to call this "direct" or "DP Indirect"? At this point, I'll do whichever. I did call out the inconsistency in the details.

Reply:
I think this should be notated as indirect.  It is after all what the letter I stands for in the mnemonic.

from https://undisbeliever.net/snesdev/65816-opcodes.html#per-push-effective-pc-relative-indirect-address
![image](https://github.com/X16Community/x16-docs/assets/395186/8b955433-bb05-4e3c-917d-36e81e7975f5)

from https://wiki.superfamicom.org/65816-reference#pei-push-effective-indirect-address-532
![image](https://github.com/X16Community/x16-docs/assets/395186/64fc2503-8b9e-46e8-822d-305e59fc5312)

Reply:
Ok. I have changed PEA to "abs" (instead of immediate), PEI to "(dir)" and PER to "rel16" to match the canonical name of the instruction.